### PR TITLE
[Feature] Support multi-class tracking in ByteTrack

### PR DESCRIPTION
**Pull request by noahcao**
add multi-class multi-object tracking support in ByteTrack based on the `labels` segments given in detections.

It would be trivial to extend this feature to other trackers.

### CODE REVIEW

File: mmtrack/models/trackers/byte_tracker.py [hunk 1]
```diff
@@ -73,8 +73,11 @@ def init_track(self, id, obj):
         bbox = bbox_xyxy_to_cxcyah(self.tracks[id].bboxes[-1])  # size = (1, 4)
         assert bbox.ndim == 2 and bbox.shape[0] == 1
         bbox = bbox.squeeze(0).cpu().numpy()
+        label_idx = self.memo_items.index('labels')
+        obj_label = obj[label_idx]
         self.tracks[id].mean, self.tracks[id].covariance = self.kf.initiate(
             bbox)
+        self.tracks[id].label = obj_label
```
Comment (GT9505) [inline mmtrack/models/trackers/byte_tracker.py:80]: Why do you need to reset `self.tracks[id].label`?
The `self.tracks[id].label` is set in `super().init_track(id, obj)`.
I think `super().init_track(id, obj)` is always equal to `obj_label`.

**Review by GT9505 — COMMENTED** [review]
1. Does the 'multi-class supporting` be the same as ByteTrack in the manuscript?
2. It is better to verify the inference performance of ByteTrack in MOT17-half-train using the pertrained weights. It should achieve 67.7 HOTA

File: mmtrack/models/trackers/byte_tracker.py (hunk 1 above)
Comment (noahcao) [reply mmtrack/models/trackers/byte_tracker.py]: I see, I didn't notice this. I have removed this duplicated `self.tracks[id].label` in the new commit.

**Comment by noahcao** [comment]
I have verified the performance alignment before and after this commit. The performance is well aligned though it is not 67.7 HOTA but slightly lower as shown below.

image

**Comment by GT9505** [comment]
> 1. this 'multi-class supporting' does no require any modification in the manuscript as it is directly based on the `labels` which is already input to the tracker in the previous version.
> 2. I have verified the performance alignment before and after this commit. The performance is well aligned though it is not 67.7 HOTA but slightly lower as shown below.
> 
> image

Do you mean that we can achieve 67.7 before this pr, and 67.1 HOTA after this pr?
I do not get it. 
Should the behavior be unchanged when there is only one category to track?

**Comment by noahcao** [comment]
> > 1. this 'multi-class supporting' does no require any modification in the manuscript as it is directly based on the `labels` which is already input to the tracker in the previous version.
> > 2. I have verified the performance alignment before and after this commit. The performance is well aligned though it is not 67.7 HOTA but slightly lower as shown below.
> > 
> > image
> 
> Do you mean that we can achieve 67.7 HOTA before this pr, while achieve 67.1 HOTA after this pr? I do not get it. The behavior should have be unchanged when there is only one category to track, since all element in the `cate_cost` equal to 0.

I meant: when I ran the original code (before this PR), it achieved HOTA=67.1. After PR, it is still 67.1. I didn't ever reproduce the reported 67.7HOTA on my machine at the very beginning.

**Comment by GT9505** [comment]
I'll check it later since the bytetracker is not changed. 
I guess that it may due to the changes of yolox detector in mmdet.

**Review by GT9505 — APPROVED** [review]

**Comment by GT9505** [comment]
> I'll check it later since the bytetracker is not changed. I guess that it may due to the changes of yolox detector in mmdet.

using mmdet=2.19.1 and mmtrack=0.13.0 we can achieve 67.7 HOTA.
using mmdet=2.24.1 and mmtrack=0.13.0 we can achieve 67.1 HOTA.

**Pull request merged**